### ISSUE TITLE
iot-device-client 1.32.1 with edgeHub 1.1.4 prints repeated warning "A message was received with a null correlation id"

### ISSUE THREAD

**Issue by derSchtefan** [Azure/azure-iot-sdk-java#1349]
# Context

- Ubuntu 18.04
- openjdk version "11.0.12" 2021-07-20 LTS
- iot-device-client 1.32.1

## Description of the issue
We are processing messages from the iot edge edgeHub  1.1.4 like recommended in the examples:
```java
IotHubClientProtocol protocol = IotHubClientProtocol.MQTT;
log.info("Start to create client with MQTT protocol");
ModuleClient client = ModuleClient.createFromEnvironment(protocol);
log.info("Client created");
client.setMessageCallback(IotHubCallbacks.INPUT_NAME, msgCallback, client);
```

The messages are received from the Azure Industrial-IOT  OpcPublisher. 

When upgrading to iot-device-client 1.32.1, we receive the warning:
` c.m.a.s.i.d.transport.IotHubTransport    : A message was received with a null correlation id.`
logged for each received message. It floods the message log.

It did not happen in earlier versions of the SDK. (1.31.0)

Since we have no control over OpcPublisher and what kind of messages it sends, and those correlation ids are not documented, what are the implications. how can we fix this issue, or do we have to silence and ignore the warning ?

**Comment by DBY047** [comment]
I see the same things on my side but the warning message is wrong. First, the correlationId cannot be null. getCorrelationId will return an empty string if correlationId is null on the Message object. Second, the real error is that correlationCallbacks  containsKey return null for that correlationId: correlationCallbacks is empty when this occurred. I see 2 of these warnings. I added more about the message below. 

```
A message was received with a null correlation id:  
Message details: Correlation Id [c8a6f487-d075-48ef-8bcf-ed513b91b145] Message Id [80c9b4f0-56f4-4106-a26c-27ed2bc30756] Request Id [c8a6f487-d075-48ef-8bcf-ed513b91b145] Device Operation Type [DEVICE_OPERATION_TWIN_GET_RESPONSE] 

A message was received with a null correlation id:  
Message details: Correlation Id [0e70a3eb-97e1-4bfa-abcb-b2386a8d74fc] Message Id [b8568cb8-1487-45da-8da9-9e77e8992d5f] Request Id [0e70a3eb-97e1-4bfa-abcb-b2386a8d74fc] Device Operation Type [DEVICE_OPERATION_TWIN_UPDATE_REPORTED_PROPERTIES_RESPONSE]

**Comment by DBY047** [comment]
Also have this when sending a direct-method to my device.  It looks like nothing sets correlationCallbacks in the IotHubTransport but now it complains about it. Are we suppose to implement the CorrelatingMessageCallback interface?

**Comment by DBY047** [comment]
I think the handling of correlationCallbacks stuff in IotHubTransport can be improved a bit.  

1. no need to check if messageId is null getCorrelationId will return an empty string if it is and it isn't checked in addToWaitingQueue.
2. A warning have been added only in onMessageReceived. All other method of the CorrelatingMessageCallback interface have no warning if correlationId is not in the correlationCallbacks map (or null) . I think the warning can be removed.
3. There is an nullPointerException if the message.getCorrelatingMessageCallbackContext() is null. Others API in the SDK allow the context object to be null. 

What do you think?

**Comment by timtay-microsoft** [comment]
Tagging @jamdavi who has a bit more context on why we'd be checking for this correlation id

**Comment by jamdavi** [comment]
Thanks everyone, I added this as a way to correlate messages. I'll see about improving the messages so they don't spam

**Comment by derSchtefan** [comment]
@jamdavi Why is it suddenly worth a log entry if a message has no correlation id?

**Comment by jamdavi** [comment]
Thanks everyone, great feedback on this. The PR is out and should be reviewed soon.

@DBY047 
> ...the warning message is wrong...

This is true, the warning message should have said something like "The correlation callback dictionary does not have an entry for correlation id..."

> Are we suppose to implement the CorrelatingMessageCallback interface?

No, it's optional.

> I think the handling of correlationCallbacks stuff in IotHubTransport can be improved a bit.

I like all of these suggestions and implemented them in a PR. 

@derSchtefan 
> Why is it suddenly worth a log entry if a message has no correlation id?

When testing this I had reused a few log statements to follow along. It should have been removed or at the very least been set as debug or trace.
